File: k-random-bean-validation/src/test/java/io/github/krandom/validation/PatternAnnotationHandlerTest.java
package io.github.krandom.validation;

import static org.assertj.core.api.Assertions.assertThat;

import io.github.krandom.KRandom;
import jakarta.validation.ConstraintViolation;
import jakarta.validation.Validation;
import jakarta.validation.Validator;
import jakarta.validation.ValidatorFactory;
import jakarta.validation.constraints.Pattern;
import java.util.Set;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class PatternAnnotationHandlerTest {
  private KRandom kRandom;

  @BeforeEach
  void setUp() {
    kRandom = new KRandom();
  }

  @Test
  void generatedBeanShouldBeValidAccordingToValidationConstraints() {
    TestBean testBean = kRandom.nextObject(TestBean.class);

    assertThat(testBean.getTestString()).matches("^A$");
  }

  @Test
  void generatedBeanShouldBeValidUsingBeanValidationApi() {
    TestBean testBean = kRandom.nextObject(TestBean.class);

    Validator validator;
    try (ValidatorFactory validatorFactory = Validation.buildDefaultValidatorFactory()) {
      validator = validatorFactory.getValidator();
    }
    Set<ConstraintViolation<TestBean>> violations = validator.validate(testBean);

    assertThat(violations).isEmpty();
  }

  static class TestBean {
    @Pattern(regexp = "^A$")
    private String testString;

    public String getTestString() {
      return testString;
    }
  }
}

File: bean-validation/src/test/kotlin/io/github/krandom/validation/PatternAnnotationHandlerTest.kt
package io.github.krandom.validation

import io.github.krandom.KRandom
import io.kotest.matchers.collections.shouldBeEmpty
import io.kotest.matchers.string.shouldMatch
import jakarta.validation.Validation
import jakarta.validation.Validator
import jakarta.validation.constraints.Pattern
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

private const val REGEX: String = """^[A-Z]{1,2}[0-9][0-9A-Z]? ?[0-9][A-Z]{2}$"""

internal class PatternAnnotationHandlerTest {
  private lateinit var kRandom: KRandom

  @BeforeEach
  fun setUp() {
    kRandom = KRandom()
  }

  @Test
  fun generatedBeanShouldBeValidAccordingToValidationConstraints() {
    val testBean = kRandom.nextObject(TestBean::class.java)

    testBean.testString shouldMatch REGEX
  }

  @Test
  fun generatedBeanShouldBeValidUsingBeanValidationApi() {
    val testBean = kRandom.nextObject(TestBean::class.java)

    val validator: Validator
    Validation.buildDefaultValidatorFactory().use { validatorFactory ->
      validator = validatorFactory.validator
    }
    val violations = validator.validate(testBean)

    violations.shouldBeEmpty()
  }

  internal data class TestBean(@field:Pattern(regexp = REGEX) val testString: String)
}
